build/lib/src/generate/expected_outputs.dart

```diff
@@ -7,22 +7,148 @@ import '../builder/builder.dart';
-  return AssetId(
-      assetId.package,
-      path.replaceRange(
-          path.length - oldExtension.length, path.length, newExtension));
+// We can safely cache parsed build extensions for each builder since build
+// extensions are required to not change for a builder.
+final _parsedInputs = Expando<List<_ParsedBuildOutputs>>();
+
+/// Extensions on [Builder] describing expected outputs.
+extension BuildOutputExtensions on Builder {
+  /// Whether this builder is expected to output assets when running on [input].
+  ///
+  /// This will be `true` iff its [expectedOutputs] is not empty, but may be
+  /// more efficient to compute.
+  bool hasOutputFor(AssetId input) {
+    return _parsedExtensions.any((e) => e.hasAnyOutputFor(input));
+  }
+
+  List<_ParsedBuildOutputs> get _parsedExtensions {
+    return _parsedInputs[this] ??= [
+      for (final entry in buildExtensions.entries)
+        _ParsedBuildOutputs.parse(this, entry.key, entry.value),
+    ];
+  }
+}
+
+extension on AssetId {
+  /// Replaces the last [matchedSuffix] characters with [newExtension].
+  AssetId replaceSuffix(int matchedSuffix, String newExtension) {
+    return AssetId(
+        package, path.substring(0, path.length - matchedSuffix) + newExtension);
+  }
+}
+
+abstract class _ParsedBuildOutputs {
+  static final RegExp _captureGroup = RegExp('{{}}');
+
+  _ParsedBuildOutputs();
+
+  factory _ParsedBuildOutputs.parse(
+      Builder builder, String input, List<String> outputs) {
+    final groups = _captureGroup.allMatches(input).iterator;
+
+    if (!groups.moveNext()) {
+      // The input does not contain a capture group, so we should simply match
+      // all assets whose paths ends with the desired input.
+      return _SuffixBuildOutputs(input, outputs);
+    }
+
+    final match = groups.current;
+
+    if (groups.moveNext()) {
+      throw ArgumentError(
+        'The builder `$builder` declares an input "$input" which contains '
+        'multiple groups ("{{}}"). This is not supported.',
+      );
+    }
+
+    // When using a capture group in the build input, it must also be used in
+    // every output to ensure outputs have unique names.
+    for (final output in outputs) {
+      if (_captureGroup.allMatches(output).length != 1) {
+        throw ArgumentError(
+          'The builder `$builder` declares an input "$input" using a capture '
+          'group. It is required that all of its outputs also refer to that '
+          'capture group exactly once. However, "$output" does not.',
+        );
+      }
+    }
+
+    final regexBuffer = StringBuffer();
+    // Start writing the regex up to the capture group
+    if (input.startsWith('^')) {
+      // Build inputs starting with `^` start matching at the beginning of the
+      // input path.
+      regexBuffer
+        ..write('^')
+        ..write(RegExp.escape(input.substring(1, match.start)));
+    } else {
+      regexBuffer.write(RegExp.escape(input.substring(0, match.start)));
+    }
```

Comment:
What cases is this conditional impacting the end result?  We know the `.substring(1` call is always removing exactly `^`, and that is what we replace it with.

Can this be

```dart
final regexBuffer = StringBuffer(input.substring(0, match.start));
```

Reply:
What if there are other characters from `1` to `match.start` that might need escaping? If `input` was `^.hidden/{{}}`, I think we'd want to generate a `^\.hidden/(.*)` regex, right? We can't escape all of it either since we need the `^` at the beginning.

Reply:
Oh right, I completely overlooked the escaping. 😳